pkg/ingester/user_metrics_metadata.go

```diff
@@ -86,15 +87,32 @@ func (mm *userMetricsMetadata) purge(deadline time.Time) {
 	mm.metrics.memMetadataRemovedTotal.WithLabelValues(mm.userID).Add(float64(deleted))
 }
 
-func (mm *userMetricsMetadata) toClientMetadata() []*mimirpb.MetricMetadata {
+func (mm *userMetricsMetadata) toClientMetadata(req *client.MetricsMetadataRequest) []*mimirpb.MetricMetadata {
 	mm.mtx.RLock()
 	defer mm.mtx.RUnlock()
-	r := make([]*mimirpb.MetricMetadata, 0, len(mm.metricToMetadata))
-	for _, set := range mm.metricToMetadata {
+	rCap := int32(len(mm.metricToMetadata))
+	if req.Limit >= 0 && req.Limit < rCap {
```

Comment:
Is 0 a valid limit? What's the purpose of sending a request with limit = 0?

Reply:
I checked Prometheus and it actually support a value of 0, returning no metadata at all. Not sure whether it's intentional or not, but let's keep Mimir consistent with Prometheus for now. However, I suggest you to open an issue in Prometheus to double check if that's intentional.

Reply:
It also looks inconsistent with `LimitPerMetric` where a value of 0 means unlimited.

Reply:
https://github.com/prometheus/prometheus/issues/12862